llvm/lib/Target/X86/X86AsmPrinter.cpp

```diff
@@ -55,6 +55,116 @@ X86AsmPrinter::X86AsmPrinter(TargetMachine &TM,
 // Primitive Helper Functions.
 //===----------------------------------------------------------------------===//
 
+const TargetInstrInfo *TII;
+
+/// Clear any mappings that map to the given register.
+void clearRhs(Register Reg, std::map<Register, int64_t> &SpillMap) {
+  for (auto I = SpillMap.begin(); I != SpillMap.end();) {
+    if (I->second == Reg) {
+      I = SpillMap.erase(I);
+    } else {
+      ++I;
+    }
+  }
+}
+
+/// Given a MachineBasicBlock, analyse its instructions to build a mapping of
+/// where duplicate values live. This can be either in another register or on
+/// the stack. Since registers are always positive and stack offsets negative,
+/// we can store both in the same map while still being able to distinguish
+/// them. This allows us to keep the changes in Stackmaps.cpp to a minimum and
+/// avoids having to alter the stackmap format.
+void processInstructions(
+    const MachineBasicBlock *MBB,
+    std::map<Register, int64_t> &SpillMap,
+    std::map<const MachineInstr *, std::map<Register, int64_t>> &StackmapSpillMaps
+  ) {
+  for (const MachineInstr &Instr : MBB->instrs()) {
+
+    // At each stackmap call, save the current mapping so it can later be
+    // encoded in the stackmap when it is lowered.
+    if (Instr.getOpcode() == TargetOpcode::STACKMAP) {
+      StackmapSpillMaps[&Instr] = SpillMap;
+      continue;
+    }
+
+    // Copying a value from one register B to another A, creates a mapping from
+    // A to B. If A is tracked by the stackmap, then B will also be tracked and
+    // assigned the same value during deoptimisation.
+    if (Instr.isMoveReg()) {
+      const MachineOperand Lhs = Instr.getOperand(0);
+      const MachineOperand Rhs = Instr.getOperand(1);
+      assert(Lhs.isReg() && "Is register.");
+      assert(Rhs.isReg() && "Is register.");
+      SpillMap[Lhs.getReg()] = Rhs.getReg();
+      // Reassigning a new value to Lhs means any mappings to Lhs are now void
+      // and need to be removed.
+      clearRhs(Lhs.getReg(), SpillMap);
+      continue;
+    }
+
+    int FI;
+    // A value from a register was spilled to the stack. Create a mapping from
+    // that register to the stack offset.
+    if (TII->isStoreToStackSlotPostFE(Instr, FI)) {
```

Comment:
Maybe there's a C++-ism I'm not aware of, but this looks to me like we're passing an uninitialised variable `FI` to a function.

Reply:
Is `FI` pass by reference?

Reply:
Found it, yes:

https://llvm.org/doxygen/classllvm_1_1X86InstrInfo.html#a2da9f36c923f6d843b1bc13c35b75c16

So I guess we are passing in an uninitialised variable for the function to populate?

Reply:
Then it should, I believe, be `&FI`, or does that conversion magically happen in C++?

Reply:
Yes, it's a reference. `isStoreToStackSlotPostFE` writes the frame index into `FE`, though we are not doing anything with it here.

Reply:
In C, `FI` means "the value of FI" and `&FI` means "a reference to the variable CI". I don't know if C++ automagically converts between the two, but either way, I would guess that writing `&FI` here is at least (probably) clearer and (possibly) also fixes a bug?

Reply:
> Then it should, I believe, be `&FI`, or does that conversion magically happen in C++?

I'm no C++ expert, but I assumed it does it automatically. I might have even tried doing `&F` and the compiler complained. But I can't remember.

Reply:
If I do `&FI` I get `/home/lukasd/research/ykllvm/llvm/lib/Target/X86/X86AsmPrinter.cpp:109:46: error: invalid conversion from ‘int*’ to ‘int’ [-fpermissive]`. I did seem to remember `&FI` is the first thing I tried for the same reasons you listed.

Reply:
I'm almost certainly wrong on this so will resolve.